### ISSUE TITLE
【模型预训练】如何训练一个模型

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#40]
# 什么是神经网络

神经网络由处理数据的层和模块构成：

<img width="513" alt="image" src="https://github.com/QingyaFan/blog/assets/10420579/47bd190d-8aff-41e0-936b-74391150cc00">

神经网络的特点：
- 可以自动地从数据中学习到合适的权重参数，权重是在一个层内各个神经元的影响程度。如上图中 w11,w12,w21,w22,23；
- 说一个网络有几层，一般是指网络中有权重的层的数量，对于上图来讲，是 2 层。

虽然上图中我们看到每一层是多个分散的神经元，但实际上每一层实际上是一个 module，我们给它 input，它吐出 output，当前 input 和 output 类型一般是 tensor 类型。从这个角度看，神经网络实际上是多个接受和输出tensor数据的多个 module 串联而成的。

## 如何从数据中学习到权重参数呢

https://pytorch.org/tutorials/beginner/basics/buildmodel_tutorial.html#define-the-class
```
import os
import torch
from torch import nn
from torch.utils.data import dataloader
from torchvision import datasets, transforms

device = (
    "cuda"
    if torch.cuda.is_available()
    else "mps"
    if torch.backends.mps.is_available()
    else "cpu"
)
print(f"Using {device} device")

class NeuralNetwork(nn.Module):
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28*28, 512),
            nn.ReLU(),
            nn.Linear(512, 512),
            nn.ReLU(),
            nn.Linear(512, 10),
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        return logits
    

model = NeuralNetwork().to(device)

print(model)

X = torch.rand(1, 28, 28, device=device)
# Logits are the outputs of a neural network 
# before the activation function is applied.
logits = model(X)
pred_probab = nn.Softmax(dim=1)(logits)
y_pred = pred_probab.argmax(1)
print(f"Predicted class: {y_pred}")
```

代码中的 softmax 的作用：
> Applies the Softmax function to an n-dimensional input Tensor rescaling them so that the elements of the n-dimensional output Tensor lie in the range [0,1] and sum to 1.

# 使用 PyTorch 训练一个神经网络

## 前置需要了解的概念：

- Loss Function（损失函数），表示神经网络的恶劣程度的指标，即当前的神经网络对监督数据在多大程度上不拟合、不一致。训练的目的就是以损失函数为基准，找出能使它的值达到最小的权重参数。常用的有均方误差、交叉熵误差
- Optimizer（优化器），参数寻优（参数指`权重`和`偏置`），损失函数取最小值时的参数就是`最优参数`。但是，一般损失函数复杂、参数空间庞大，我们不知道它在何处能取得最小值，而通过使用梯度来寻找函数最小值（或者尽可能小的值）的方法就是`梯度法`。梯度会指向各点处的函数值降低的方向，更严格的说，梯度指示的方向是各点处函数值减小最多的方向
- Embeding（嵌入）：是一种将高维度的离散数据（如用户、物品）映射到低维度连续向量空间的方法，可以减低计算复杂度，在机器学习和深度学习中非常常见，尤其是在自然语音处理（NLP）、推荐系统和计算机视觉中
- Expand Embeding（扩展嵌入），扩展嵌入是对基础嵌入的增强，例如基础嵌入是（用户、物品），扩展嵌入可以表示额外的上下文信息，如用户的历史行为、物品的类别特征、事件和位置等。当然，可以增大基础嵌入的维度达到同样的效果，但基于模型扩展性、计算和内存效率、嵌入的复用等考虑，一般用扩嵌入
- ParameterServer（PS），分布式机器学习中，参数服务器用于管理和共享模型参数，其基本思想是将模型参数存储在一个或多个中央服务器上，并通过网络将这些参数共享给参与训练的各个计算节点。每个计算节点可以从参数服务器中获取当前模型参数，并将计算结果返回给参数服务器进行更新。 [^1] 利用 PyTorch 等框架提供的分布式能力可以实现典型的 PS。[^2]

## 训练一个模型的步骤

1. 定义网络结构
7. 定义数据
8. 定义损失函数和优化器
9. 训练循环
10. 保存为模型文件

### 1. 定义网络结构

例如，我们继承 pytorch 的 nn.Module 定义一个模型：

```
class GarmentClassifier(nn.Module):
    def __init__(self, *args, **kwargs) -> None:
        super(GarmentClassifier, self).__init__(*args, **kwargs)
        self.conv1 = nn.Conv2d(1, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 4 * 4, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 16 * 4 * 4)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x
```

模型的必备要素：
1. 定义一些层，例如卷积层、池化层、全连接层
4. 数据在这些层之间如何传递，即前向传播函数

### 2. 定义数据

```
# create datasets for training & validation
training_set = torchvision.datasets.FashionMNIST(
    './data',
    train=True,
    transform=transform,
    download=True
)
validation_set = torchvision.datasets.FashionMNIST(
    './data',
    train=False,
    transform=transform,
    download=True
)

# create data loaders for our datasets;
# shuffle for training, not for validation
training_loader = torch.utils.data.DataLoader(
    training_set,
    batch_size=4,
    shuffle=True
)
validation_loader = torch.utils.data.DataLoader(
    validation_set,
    batch_size=4,
    shuffle=False
)
```

### 3. 定义损失函数和优化器

```
loss_fn = torch.nn.CrossEntropyLoss()
dummy_outputs = torch.rand(4, 10)
dummy_labels = torch.tensor([1,5,3,7])
loss = loss_fn(dummy_outputs, dummy_labels)
```

用了交叉熵损失函数，通过将模型的输出和真实标签作为参数传递给损失函数，计算得到了模型的预测结果与真实标签之间的损失值。损失值越小，模型的性能越好。

```
optimizer = torch.optim.SGD(model.parameters(), lr=0.001, momentum=0.9)
```

使用了随机梯度下降（SGD）优化器来优化模型的参数：
- 优化器根据 model.parameters() 返回的参数来更新模型的权重和偏置，
- lr=0.001 表示每次更新时参数沿着梯度方向变化的步长，这个参数控制了模型参数更新的速度，
- momentum=0.9 用于加速 SGD 在参数空间中的收敛

### 4. 训练循环 

todo

### 5. 保存为模型文件

todo

## 分布式训练

PyTorch中，Distributed Data Parallel（DDP）是常用且推荐的分布式训练方式，适用于多GPU和多节点的分布式训练场景。DDP的核心概念：

1. 数据并行：DDP 在多个设备（例如GPU）上并行训练同一个模型，每个设备有一份独立的模型副本，处理不同的数据子集
4. 同步更新：在每个训练步骤结束时，DDP会通过 AllReduce 操作同步所有模型副本的梯度。每个副本计算自己的一部分梯度，然后所有副本的梯度会被求和，得到全局梯度，用于更新模型参数
5. 高效通信：DDP 使用高效的通信后端（例如NCCL、Gloo）进行梯度同步，极大地减少通信开销，特别是在多 GPU 环境中

TensorFlow中，ParameterServerStrategy 适用于大规模分布式训练，尤其是需要复杂的计算和存储架构的任务。ParameterServerStrategy 是一种经典的参数服务器架构，用于在多个参数服务器（ps）和多个工作节点（worker）之间分布模型参数和训练任务：

- 工作原理：参数服务器负责存储模型参数，而工作节点负责计算梯度并更新参数服务器上的参数。适用于需要跨多个机器分布大量计算任务的情况。

## 分布式推理

todo：当一张卡上无法容纳模型时，如何拆分模型到多张卡进行联合推理

[^1]: https://www.cnblogs.com/N3ptune/p/17437320.html
[^2]: https://pytorch.org/tutorials/intermediate/rpc_param_server_tutorial.html
